### PR TITLE
Add malloc() and related functions to minimal libc

### PR DESCRIPTION
**Pull request by tyrantlizard77**
The immediate need for this is to support getaddrinfo(), which requires a memory allocation.

Uses existing sys_mem_pool implementation.

### CODE REVIEW

File: lib/libc/minimal/source/stdlib/calloc.c
Comment (lpereira) [inline lib/libc/minimal/source/stdlib/calloc.c]: These files should have the license boilerplate/copyright information.

File: lib/libc/minimal/source/stdlib/calloc.c
Comment (lpereira) [inline lib/libc/minimal/source/stdlib/calloc.c]: Indeed.  __builtin_mul_overflow() is cheap and we should use it. :)

File: lib/libc/minimal/source/stdlib/malloc.c
Comment (lpereira) [inline lib/libc/minimal/source/stdlib/malloc.c]: Nitpick: Indentation seems a little off in this file

File: lib/libc/minimal/source/stdlib/realloc.c
Comment (lpereira) [inline lib/libc/minimal/source/stdlib/realloc.c]: We can improve this later, but... `if (newptr != NULL) { memcpy(...); free(ptr); }` is needed here.

**Comment by andrewboie** [comment]
@lpereira this is a vanilla import of the baselib code.
I'd like to get TSC approval that this can actually go in, before I start making too many changes to the baselib malloc implementation, as it could very well be rejected and I have to pull in some other malloc implementation or write my own

File: lib/libc/minimal/source/stdlib/calloc.c
Comment (andrewboie) [reply lib/libc/minimal/source/stdlib/calloc.c]: added.

**Comment by pfalcon** [comment]
>  the baselib malloc implementation, as it could very well be rejected and I have to pull in some other malloc implementation or write my own

Not sure about TSC, but even myself the mere developer would be interested to know why malloc implementation was chosen from https://github.com/PetteriAimonen/Baselibc, which other alternatives were considered?

**Comment by andrewboie** [comment]
@pfalcon if you have another compact malloc implementation with a compatible license you'd rather we use please link it.

**Review by andyross — APPROVED** [review]
Would it be worth adding a kconfig that enables the heap and have getaddrinfo() and other future stuff depends/select it via some mechanism?

The problem with adding a feature like this is that once it's available for use, everything is going to start using it and the "Zephyr works without a heap" promise becomes pretty empty, as only a mostly-academic subset of the OS will be heap-free.   At least if you have to turn it on, we can easily detect when we're adding new dependencies.

No complaints with the code itself.

**Review by nashif — DISMISSED** [review]
blocking until we get approval from TSC/Board with regard to the license.

**Comment by andrewboie** [comment]
> FWIW: you could wrap k_malloc() as an alternative. It's not a great fit for a general purpose allocator (though frankly this one doesn't look like great shakes either), but is already in-tree.

@andyross no, you can't. This needs to be usable from user mode and k_mem_pools and the system heap are unsuitable.

**Comment by andyross** [comment]
Didn't we sort out userspace support for mempool, though?  Seems like that might be a better use of our time than importing a foreign allocator just for getaddrbyname().

**Comment by andrewboie** [comment]
> Didn't we sort out userspace support for mempool, though? Seems like that might be a better use of our time than importing a foreign allocator just for getaddrbyname().

I could use sys_mem_pool but it's so clumsy to use.
The only supported size values for CONFIG_HEAP_MEM_POOL_SIZE are 256, 1024, 4K, and 16K
I am loath to introduce a malloc to the C library that is so limited.
This thing lets you add whatever size arena you want.

**Comment by andyross** [comment]
Huh, yeah, that's a stale comment in the Kconfig.  In fact K_MEM_POOL_DEFINE works with arbitrary heap sizes, that was one of the goals of the rewrite.  It's a power-of-4 allocator still, so it's fairly wasteful for generic malloc usage in some workloads.   But really I think a mempool should do us pretty well.

**Comment by andrewboie** [comment]
>  K_MEM_POOL_DEFINE works with arbitrary heap sizes, that was one of the goals of the rewrite

If that's the case then sys_mem_pool (mostly same implementation) should work, I'll revise this PR to use it.

File: lib/libc/minimal/source/stdlib/malloc.c
Comment (andyross) [inline lib/libc/minimal/source/stdlib/malloc.c]: FWIW: 64 is a pretty chunky minium size.  The overhead for allowing smaller blocks isn't super high (amortized it's like 12-ish bytes per power of two level, I think?), so you could probably go lower if you wanted.  The absolute minimum is 8.

File: lib/libc/minimal/source/stdlib/malloc.c
Comment (andyross) [inline lib/libc/minimal/source/stdlib/malloc.c]: Would be cleaner to expose the needed bits from the mempool API directly, like "sys_mem_pool_get_block(void *ptr)" or whatever.  Probably not a big deal in internal code like this.

File: lib/libc/minimal/source/stdlib/malloc.c
Comment (andyross) [inline lib/libc/minimal/source/stdlib/malloc.c]: Ditto.  sys_mem_pool_block_size(void *) would be prettier.

File: lib/libc/minimal/include/stdlib.h
Comment (andyross) [inline lib/libc/minimal/include/stdlib.h]: FWIW: if the arena size is zero (the default, which I like!), all the allocation calls will soft fail at runtime.  Might be more helpful (albeit less standards compliant) to #if out the declarations entirely in that case so users don't get confused about why their code is failing.

**Review by andyross — APPROVED** [review]
Much smaller.  Nice.  Random nitpicking.

File: lib/libc/minimal/source/stdlib/malloc.c
Comment (ceolin) [inline lib/libc/minimal/source/stdlib/malloc.c]: Would be nice some debug info here informing that memory arena's size was not set.  This can avoid people digging into the code to see why memory allocation is failing.

File: lib/libc/minimal/source/stdlib/malloc.c
Comment (andrewboie) [reply lib/libc/minimal/source/stdlib/malloc.c]: Done.

File: lib/libc/minimal/source/stdlib/malloc.c
Comment (nashif) [inline lib/libc/minimal/source/stdlib/malloc.c]: file missing header and license

File: lib/libc/Kconfig
Comment (nashif) [inline lib/libc/Kconfig]: if this is going to be minimal libc only, does it make sense to put this in the name as a minimal libc feature?

File: lib/libc/minimal/source/stdlib/malloc.c
Comment (andrewboie) [reply lib/libc/minimal/source/stdlib/malloc.c]: done

**Review by nashif — APPROVED** [review]
We do need tests here, so hopefully we will have them by 1.13 (@kumarvikash1 )

**Pull request merged**
